metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,285 @@
+
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+import java.io.PrintWriter
+import java.io.StringWriter
+import java.net.URI
+import java.net.URLDecoder
+import java.nio.charset.StandardCharsets
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+import scala.meta.cli.Reporter
+import scala.meta.internal.builds.ShellRunner
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metap.Main
+import scala.meta.internal.mtags.SemanticdbClasspath
+import scala.meta.io.AbsolutePath
+import scala.meta.metap.Format
+import scala.meta.metap.Settings
+import scala.util.Properties
+
+final class FileDecoderProvider(
+    workspace: AbsolutePath,
+    buildTargets: BuildTargets,
+    userConfig: () => UserConfiguration,
+    shellRunner: ShellRunner
+)(implicit ec: ExecutionContext) {
+
+  private sealed abstract class Error(val message: String)
+  private sealed abstract class ErrorWithException(
+      message: String,
+      val ex: Throwable
+  ) extends Error(message)
+  private case class InvalidURI(uri: String, override val ex: Throwable)
+      extends ErrorWithException(s"Invalid URI `$uri`", ex)
+  private case class InvalidScheme(uri: URI)
+      extends Error(s"Invalid scheme `${uri.getScheme}`")
+  private case class InvalidQuery(uri: URI)
+      extends Error(s"Invalid query `${uri.getQuery}`")
+  private case class InvalidQueryParams(
+      params: Array[Array[String]],
+      override val ex: Throwable
+  ) extends ErrorWithException(
+        s"Invalid parameters `${params.map(_.mkString("=")).mkString(",")}`",
+        ex
+      )
+  private case class MissingDecoder(uri: URI)
+      extends Error(s"Missing `decoder` argument `$uri`")
+  private case class MissingOriginalURI(uri: URI)
+      extends Error(s"Missing `originalURI` argument `$uri`")
+  private case class InvalidURIPath(path: String, override val ex: Throwable)
+      extends ErrorWithException(s"Invalid URI path `$path`", ex)
+  private case class FileMissing(path: AbsolutePath)
+      extends Error(s"File missing `$path`")
+  private case class InvalidDecoder(decoder: String)
+      extends Error(s"Invalid `$decoder`")
+  private case class FileToDecodeMissing(path: AbsolutePath)
+      extends Error(s"""|No file to decode for $path
+                        |Is Metals still starting?
+                        |Is the file compiled?
+                        |""".stripMargin)
+  private case class DecoderError(
+      path: AbsolutePath,
+      override val ex: Throwable
+  ) extends ErrorWithException(s"Error decoding file `$path`", ex)
+
+  private case class Decoder(
+      findFile: AbsolutePath => Option[AbsolutePath],
+      decode: AbsolutePath => Future[Either[Error, String]]
+  )
+
+  /**
+   * URI format...
+   * metalsDecodeFile:/fileUrl?decoder=decoderName&keyValueArguments&originalURI=originalFileURI
```

Comment:
If verbose option is gone are `keyValueArguments` still necessary?

Reply:
Verbose is now making a comeback